File: progress_bar.py
#!/opt/homebrew/bin/python3

import sys
import time
class ProgressBar:
	def __init__(self, total: int, bar_length=20, done_message="Finished!", update_interval=1):
		self.total = total
		self.bar_length = bar_length
		self.done_message = done_message
		self.update_interval = update_interval
		self.current = 0

		print("\n\x1b[32m----Starting Program----\x1b[0m\n")

	def update(self, current: int):
		self.current = current
		fraction = self.current / self.total

		arrow = int(fraction * self.bar_length - 1) * '-' + '>'
		padding = (self.bar_length - len(arrow)) * ' '

		if self.current == self.total:
			ending = f'\n\n{self.done_message}\n\n'
		else:
			ending = '\r'

		completed = '\x1b[32m' if self.current == self.total else '\x1b[0m'

		print(
		f"{completed}Progress: [{arrow}{padding}] {int(fraction*100)}%",
		end=ending
		)

	def increment(self, step=1):
		self.update(self.current + step)

# Example usage
total_items = 100
progress_bar = ProgressBar(total_items)

for i in range(total_items + 1):
	try:
		progress_bar.increment()
		time.sleep(0.1)  # Simulate some processing time
	except KeyboardInterrupt:
		sys.stdout.write("\x1b[2K")
		sys.stdout.write("\x1b[31mCancelled...\x1b[0m\n\n")
		sys.stdout.flush()
		exit()

File: progress_bar.cpp
#include <iostream>
#include <string>
#include <thread>
#include <chrono>

using namespace std;

class ProgressBar {
	private:
		int total;
		int current;
		int bar_length;
		string message;

	public:
		ProgressBar(int total, int bar_length = 20, const string &message = "Finished!"): total(total), current(0), bar_length(bar_length), message(message) {
			std::cout << "\n\x1b[32m----Starting Program----\x1b[0m\n" << std::endl;
    }

		void update(int current) {
			double fraction = static_cast<double>(current) / total;

			string arrow;
			for (int i = 0; i < static_cast<int>(fraction * bar_length) - 1; ++i){
				arrow += "-";
			}
			arrow += ">";

			string padding;
			for (int i = 0; i < bar_length - arrow.size(); ++i){
				padding += " ";
			}

			string ending = current == total ? "\n\n" + message + "\n\n" : "\r";
			string completed = current == total ? "\x1b[32m" : "\x1b[0m";

			cout << completed << "Progress: [" << arrow << padding << "] " << fraction * 100 << "%" << ending;
		}

		void increment(int step = 1) {
			current += step;
			update(current);
		}

		void mark_as_completed() {
			current = total;
			update(current);
		}

		void simulateFileDownload() {
			auto startTime = std::chrono::high_resolution_clock::now();
			while (current < total) {
				auto currentTime = std::chrono::high_resolution_clock::now();
				auto elapsed = std::chrono::duration_cast<std::chrono::seconds>(currentTime - startTime).count();
				if (elapsed >= 1) { // Update every 1 second
					increment(10);
					std::cout << std::flush; // Flush the output
					startTime = currentTime;
					startTime = currentTime;
				}
				// Simulate downloading work (replace this with actual work)
			}
		}
};

